### PR TITLE
ONNX Export Support for CrossEntropyLoss

### PR DESCRIPTION
**Pull request by KsenijaS**
Add ONNX export support for torch.nn.CrossEntropyLoss.

This PR makes following changes:
1. Updates nll_loss export
2. Makes a post pass for SoftmaxCrossEntropy

### CODE REVIEW

File: caffe2/python/onnx/tests/onnx_backend_test.py
Comment (lara-hdr) [inline caffe2/python/onnx/tests/onnx_backend_test.py]: remove print

**Review by lara-hdr — APPROVED** [review]
Please rebase to retrigger the tests

**Comment by KsenijaS** [comment]
@houseroad can you please review. CI failure is not related to my PR

File: test/onnx/test_operators.py
Comment (lara-hdr) [inline test/onnx/test_operators.py:880]: can we have one 3d-mean, one 3d-none and one 4d test? 
they all pass by different branches in the postpass

File: torch/onnx/utils.py
Comment (lara-hdr) [inline torch/onnx/utils.py:140]: nit: keep the line

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (lara-hdr) [inline torch/csrc/jit/passes/onnx/peephole.cpp:873]: This is a complex function. We need to keep comments explaining clearly each of the passes.

Explain that this pass fuses logSoftmax and NegativeLogLikelihood into SoftmaxCrossEntropy.
Also for each of the section explain why they are required (2d-3d-4d none reduction etc)

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (lara-hdr) [inline torch/csrc/jit/passes/onnx/peephole.cpp:950]: comment : this is required for the optional weight input

**Review by lara-hdr — CHANGES_REQUESTED** [review]
Since ONNX's spec changed for SoftmaxCrossEntropy and NegativeLogLikelihood could you update ONNX to the latest version?

File: test/onnx/test_operators.py
Comment (KsenijaS) [reply test/onnx/test_operators.py:880]: Should we also have tests that have inputs bigger than 4d, like 5d or 6d?

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/peephole.cpp:877]: It will help us understand the pass if you provide some examples of the graph before/after the pass.

Comment (KsenijaS) [reply torch/csrc/jit/passes/onnx/peephole.cpp:877]: before the pass:
**input (2D):**
graph(%input : Float(3, 5),
      %target : Long(3)):
 %4 : Float(3, 5) = onnx::LogSoftmax[axis=1](%input)
  %8 : Float(3) = onnx::NegativeLogLikelihoodLoss[reduction="none"]
**input (4D):**
graph(%input : Float(3, 5, 2, 7),
      %target : Long(3, 2, 7)):
  %4 : Tensor = onnx::Transpose[perm=[0, 3, 2, 1]](%input)
  %5 : Tensor = onnx::LogSoftmax[axis=3](%4)
  %6 : Float(3, 5, 2, 7) = onnx::Transpose[perm=[0, 3, 2, 1]](%5) 
  %10 : Float(3, 2, 7) = onnx::NegativeLogLikelihoodLoss[reduction="none"](%6, %target) 
  return (%10)
**input (3D and >4D):**
_reduction = mean or sum_
graph(%input : Float(3, 5, 2),                                                                                                                                                                                                                                                                                                                                                                                                                   %target.1 : Long(3, 2)):                                                                                                                                                                                                                                                                                                                                                                                                               %4 : Tensor = onnx::Transpose[perm=[0, 2, 1]](%input)                                                                                                                                                                                                                                                                                                                                                                                      %5 : Tensor = onnx::LogSoftmax[axis=2](%4)                                                                                                                                                                                                                                                                                                                                                                                                 %6 : Float(3, 5, 2) = onnx::Transpose[perm=[0, 2, 1]](%5)                                                                                                                                                                                                                                                                                 %8 : Tensor = onnx::Shape(%6)                                                                                                                                                                                                                                                                                                                                                                                                              %10 : Tensor = onnx::Constant[value={0}]()                                                                                                                                                                                                                                                                                                                                                                                                 %11 : Long() = onnx::Gather[axis=0](%8, %10)                                                                                                                                                                                                                                                                                             %13 : Tensor = onnx::Shape(%6)                                                                                                                                                                                                                                                                                                                                                                                                             %15 : Tensor = onnx::Constant[value={1}]()                                                                                                                                                                                                                                                                                                                                                                                                 %16 : Long() = onnx::Gather[axis=0](%13, %15)                                                                                                                                                                                                                                                                                             %19 : Long() = onnx::Constant[value={1}]()                                                                                                                                                                                                                                                                                                                                                                                                 %20 : Long() = onnx::Constant[value={-1}]()                                                                                                                                                                                                                                                                                                                                                                                                %21 : int[] = prim::ListConstruct(%11, %16, %19, %20)                                                                                                                                                                                                                                                                                                                                                                                      %22 : Float(3, 5, 1, 2) = onnx::Reshape(%6, %21)                                                                                                                                                                                                                                                                                           %23 : Long() = onnx::Constant[value={1}]()                                                                                                                                                                                                                                                                                                                                                                                                 %24 : Long() = onnx::Constant[value={-1}]()                                                                                                                                                                                                                                                                                                                                                                                                %25 : int[] = prim::ListConstruct(%11, %23, %24)                                                                                                                                                                                                                                                                                                                                                                                           %26 : Long(3, 1, 2) = onnx::Reshape(%target.1, %25)                                                                                                                                                                                                                                                                                       %30 : Float() = onnx::NegativeLogLikelihoodLoss[reduction="sum"](%22, %26)                                                                                                                                                                                                                                                                return (%30)

 _reduction=none_
graph(%input : Float(3, 5, 2),
      %target.1 : Long(3, 2)):
  %4 : Tensor = onnx::Transpose[perm=[0, 2, 1]](%input)
  %5 : Tensor = onnx::LogSoftmax[axis=2](%4)
  %6 : Float(3, 5, 2) = onnx::Transpose[perm=[0, 2, 1]](%5)
  %8 : Tensor = onnx::Shape(%6)
  %10 : Tensor = onnx::Constant[value={0}]()
  %11 : Long() = onnx::Gather[axis=0](%8, %10) 
  %13 : Tensor = onnx::Shape(%6)
  %15 : Tensor = onnx::Constant[value={1}]()
  %16 : Long() = onnx::Gather[axis=0](%13, %15) 
  %18 : Tensor = onnx::Shape(%6)
  %20 : Tensor = onnx::Constant[value={2}]()
  %21 : Long() = onnx::Gather[axis=0](%18, %20) 
  %24 : Long() = onnx::Constant[value={1}]()
  %25 : Long() = onnx::Constant[value={-1}]()
  %26 : int[] = prim::ListConstruct(%11, %16, %24, %25)
  %27 : Float(3, 5, 1, 2) = onnx::Reshape(%6, %26) 
  %28 : Long() = onnx::Constant[value={1}]()
  %29 : Long() = onnx::Constant[value={-1}]()
  %30 : int[] = prim::ListConstruct(%11, %28, %29)
  %31 : Long(3, 1, 2) = onnx::Reshape(%target.1, %30)
  %35 : Float(3, 1, 2) = onnx::NegativeLogLikelihoodLoss[reduction="none"](%27, %31) 
  %36 : int[] = prim::ListConstruct(%11, %21)
  %37 : Float(3, 2) = onnx::Reshape(%35, %36)
  return (%37)

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/peephole.cpp:916]: And after removing code that are deleting stuff, this part that creating softmax crossentropy node separately can be refactored.

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/peephole.cpp:923]: This is not a valid condition to test `reduction=none`, directly check if reduction attribute in origNllLossNode is "none".

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/peephole.cpp:926]: I'm also thinking if we should make these if conditions into asserts. These exact patterns are expected. If they are not observed, that means something must been wrong.

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/peephole.cpp:955]: break or continue?

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/peephole.cpp:891]: nit: this line `origNllLossNode = *it;` seems to appear in every if-else block.

**Review by BowenBao — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (KsenijaS) [reply torch/csrc/jit/passes/onnx/peephole.cpp:899]: We need to remove all the inputs of the node so that input nodes can be deleted. Before deletion there is a check node->outputs[I].uses().empty()

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (KsenijaS) [reply torch/csrc/jit/passes/onnx/peephole.cpp:955]: Maybe continue would be more correct, I put break to note that nll loss node doesn't belong to any of three cases but that would also break out of the for loop and not check if somewhere else in the graph there is nllloss not that does match one of the cases

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/peephole.cpp:899]: I think Bowen's point is that once you delete all the uses (connections) of nodes in the subgraph in between the LogSoftmax and NLLLoss, the subgraph becomes orphan (not used), and will be deleted as part of "dce" pass. 
This is to avoid explicitly:
        deleteNodes.push_back(origLogSoftmaxNode);
        deleteNodes.push_back(logSoftmaxNode);
        deleteNodes.push_back(prev);

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/peephole.cpp:955]: I think it should be continue too.

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (KsenijaS) [reply torch/csrc/jit/passes/onnx/peephole.cpp:899]: I see, in that case I don't need function recursion_del

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/peephole.cpp:975]: Why is continue needed at the end of the block?

**Comment by neginraoof** [comment]
Please check test failures. Thanks.

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/peephole.cpp:911]: nit: use `nllloss_output` directly instead of duplicate step creating `reshape_after`?

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/peephole.cpp:846]: maybe we can move the comment to the first line inside each if-else block, instead of above each block.

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/peephole.cpp:867]: nit: why is this node called logSoftmaxNode?

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/peephole.cpp:887]: Same as above

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/peephole.cpp:895]: nit: this part of comment seems compressed, is this intended?

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (KsenijaS) [reply torch/csrc/jit/passes/onnx/peephole.cpp:904]: yes reduction should be enough

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (KsenijaS) [reply torch/csrc/jit/passes/onnx/peephole.cpp:867]: it should be called transpose

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (KsenijaS) [reply torch/csrc/jit/passes/onnx/peephole.cpp:895]: no, I will fix it

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (lara-hdr) [inline torch/csrc/jit/passes/onnx/peephole.cpp:885]: let's add a check that %26 %25 %22 %21 are of the right type as well here

Comment (KsenijaS) [reply torch/csrc/jit/passes/onnx/peephole.cpp]: I think that there needs to be a check for reshape and transpose node, since there is an assumption that they exist, but I'm not sure about other nodes.

Comment (KsenijaS) [reply torch/csrc/jit/passes/onnx/peephole.cpp]: I think there needs to be a check for reshape (%26) and transpose(%4) node since there is an assumption that they exist, I'm not sure about other ones

**Comment by KsenijaS** [comment]
@houseroad I always get build errors on the same 10 CIs: 
panic: runtime error: comparing uncomparable type tar.headerError
And it doesn't even run the build scrit, it just exits with the same error. Is there something wrong with PyTorchCI?

**Review by BowenBao — APPROVED** [review]
LGTM, thanks!

**Review by lara-hdr — APPROVED** [review]
looks good. maybe try rebasing to fix the CI issue?

**Comment by KsenijaS** [comment]
@houseroad rebasing doesn't fix the failing CIs:
All of the 10 CI fails with the same type of error:

Do NOT merge master branch into pull/34830 in environment pytorch-libtorch-linux-xenial-cuda10.2-cudnn7-py3-gcc7-build
Warning: Permanently added the RSA host key for IP address '140.82.112.3' to the list of known hosts.

Warning: Permanently added the RSA host key for IP address '140.82.114.3' to the list of known hosts.

Warning: Permanently added the RSA host key for IP address '140.82.113.3' to the list of known hosts.

Warning: Permanently added the RSA host key for IP address '140.82.112.4' to the list of known hosts.

panic: runtime error: comparing uncomparable type tar.headerError

goroutine 12 [running]:
net/http.(*Request).write(0xc4205e0200, 0x56125c06c660, 0xc4203a2fc0, 0x0, 0x0, 0x0, 0x56125c06f260, 0xc4205727a0)
	/usr/local/go/src/net/http/request.go:624 +0x71d
net/http.(*persistConn).writeLoop(0xc4204a7320)
	/usr/local/go/src/net/http/transport.go:1825 +0x1ec
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:1238 +0x981

Exited with code exit status 2

Can you please take a look

**Comment by houseroad** [comment]
Could you rebase again?

**Comment by KsenijaS** [comment]
> Could you rebase again?

I did rebase, but now I have even more failed builds

**Comment by BowenBao** [comment]
@houseroad  the errors look unrelated.

**Comment by KsenijaS** [comment]
@houseroad I noticed this PR https://github.com/pytorch/pytorch/pull/33328 had the same type of failures, and the common thing between 2 PRs is that they both update onnx submodule.

**Review by houseroad — CHANGES_REQUESTED** [review]
please don't update the submodule, I will find sometime to update it from internal.

**Comment by KsenijaS** [comment]
@houseroad some tests are failing because of missing update to ONNX submodule. Does this PR have to wait for the internal update to ONNX submodule before being merged?

**Comment by KsenijaS** [comment]
@houseroad It seems that CI failures are not caused by this PR.
I disabled the tests that were failing because of missing onnx update.
LGTM

**Comment by BowenBao** [comment]
@houseroad please have a look. @KsenijaS is separating this into two PRs. The second PR will enable CrossEntropy tests, once ONNX submodule is updated.

File: caffe2/python/onnx/tests/onnx_backend_test.py
Comment (houseroad) [inline caffe2/python/onnx/tests/onnx_backend_test.py:99]: since we don't update the submodule, we don't need to update this file.

File: test/onnx/test_operators.py
Comment (houseroad) [inline test/onnx/test_operators.py:882]: Shall we enable the tests in test_opeartor.py by setting enable_onnx_checker to False?

Comment (BowenBao) [reply test/onnx/test_operators.py:882]: The `assertONNX` also runs onnx_checker.

Comment (houseroad) [reply test/onnx/test_operators.py:882]: Shall we add a flag to disable that as well in assertONNX?

Comment (houseroad) [reply test/onnx/test_operators.py:882]: Or we should remove the expect files :-)

Comment (KsenijaS) [reply test/onnx/test_operators.py:882]: @houseroad there is no clean way to just disable onnx checker because export_to_pretty_string doesn't take enable_onnx_checker flag: https://github.com/pytorch/pytorch/blob/1d8012a624e4dbc9f66c7942e82e168707796855/torch/onnx/utils.py#L412
As of now it's only possible to disable onnx cheker for all tests or none.
Do you want expect files removed?

**Review by houseroad — APPROVED** [review]

File: torch/csrc/jit/passes/onnx/peephole.cpp
Comment (jiafatom) [inline torch/csrc/jit/passes/onnx/peephole.cpp:895]: @KsenijaS My PR https://github.com/pytorch/pytorch/pull/54569 has triggered these two ASSERT failures. May I know if these two ASSERT is necessary? Just checked the below logic, but don't see any places uses `Gather` property. Thanks

**Pull request closed without merging**
